server/src-lib/Hasura/RQL/DDL/Schema.hs

```diff
@@ -56,25 +59,46 @@ import           Hasura.Server.Utils            (matchRegex)
 data RunSQL
   = RunSQL
   { rSql                      :: Text
-  , rCascade                  :: !(Maybe Bool)
+  , rCascade                  :: !Bool
   , rCheckMetadataConsistency :: !(Maybe Bool)
+  , rReadOnly                 :: !Bool
   } deriving (Show, Eq, Lift)
-$(deriveJSON (aesonDrop 1 snakeCase){omitNothingFields=True} ''RunSQL)
+
+instance FromJSON RunSQL where
+  parseJSON (Object o) = do
+    rSql <- o .: "sql"
+    rCascade <- o .:? "cascade" .!= False
+    rCheckMetadataConsistency <- o .:? "check_metadata_consistency"
+    rReadOnly <- o .:? "read_only" .!= False
+    pure RunSQL{..}
+  parseJSON _ = fail "expecting an object for args"
+
+$(deriveToJSON (aesonDrop 1 snakeCase){omitNothingFields=True} ''RunSQL)
 
 runRunSQL :: (CacheBuildM m, UserInfoM m) => RunSQL -> m EncJSON
-runRunSQL (RunSQL t cascade mChkMDCnstcy) = do
+runRunSQL RunSQL{..} = do
   adminOnly
-  isMDChkNeeded <- maybe (isAltrDropReplace t) return mChkMDCnstcy
-  bool (execRawSQL t) (withMetadataCheck (or cascade) $ execRawSQL t) isMDChkNeeded
+  metadataCheckRequired <- case rReadOnly of
+    True -> pure False
+    False -> onNothing rCheckMetadataConsistency $ isAltrDropReplace rSql
+  bool (execRawSQL rSql) (withMetadataCheck rCascade $ execRawSQL rSql) metadataCheckRequired
   where
     execRawSQL :: (MonadTx m) => Text -> m EncJSON
     execRawSQL =
-      fmap (encJFromJValue @RunSQLRes) . liftTx . Q.multiQE rawSqlErrHandler . Q.fromText
+      fmap (encJFromJValue @RunSQLRes) . liftTx . setTransactionAccess . Q.multiQE rawSqlErrHandler . Q.fromText
       where
         rawSqlErrHandler txe =
           let e = err400 PostgresError "query execution failed"
           in e {qeInternal = Just $ toJSON txe}
 
+    setTransactionAccess tx
+      | rReadOnly = setReadOnly >> tx
+      | otherwise = setReadWrite >> tx
```

Comment:
This seems insufficient, right? Since if we have a `bulk` query with a read-only `run_sql` query followed by a non-`run_sql` query that makes modifications, the transaction access state won’t be restored to read/write.

Reply:
ufff my bad! So I will reset to `READ WRITE` at the end again?

```
    setTransactionAccess tx
      | rReadOnly = setReadOnly >> tx *> setReadWrite
      | otherwise =  tx
```

Now, I can see why you might need a nice pattern here. I don't know much about this bracket-thing you mentioned. Do you want me to look into it?

Reply:
[`Control.Exception.bracket_`](https://hackage.haskell.org/package/base-4.12.0.0/docs/Control-Exception.html#v:bracket_) is a function that implements an exception-safe region with setup/teardown actions, kind of like `try...finally` in other languages (but Haskell is a little more subtle here than other languages because Haskell has asynchronous exceptions, so it’s important to use `bracket_` to deal with everything properly). To implement this, you’d do something like

```haskell
withTransactionAccess newTxAccess m = do
  oldTxAccess <- getCurrentTxAccess
  bracket_ (setTxAccess newTxAccess) (setTxAccess oldTxAccess) m
```

The problem with `bracket_` is that it’s a non-algebraic operation, i.e. it has `IO`-typed arguments as well as an `IO` result type, so it can’t be easily lifted using `liftIO`. My fancy new [`eff`](https://github.com/hasura/eff) library has support for a safe version of `bracket_` that is not coupled to `IO`, but in the meantime, you’ll probably want to use [the version from `Control.Monad.Catch`](https://hackage.haskell.org/package/exceptions-0.10.3/docs/Control-Monad-Catch.html#v:bracket_), which provides `bracket_` via the mtl-style typeclass `MonadMask`.

Reply:
I thought a bit about this and TBH, I am not sure how to implement this or where. Don't we need something like a Monad State to keep track of the current transaction access mode?

Either way, not sure if all these changes are worth it atm. What do you think of the current hack?